### ISSUE TITLE
Unhanlded VR exception in FODICOM CSTORESCP SAMPLE 

### ISSUE THREAD

**Issue by dglecons** [fo-dicom/fo-dicom#378]
Study can be found from the below link
https://www.sendspace.com/file/wmo4jx
### Expected behavior
KPACS should be able to send attached study to FODICOM CStoreSCP sample.

KPACS version 0.9.9

### Actual behavior
CStoreSCP receives first image and reports Unhandled VR exception.

### Steps to reproduce the behavior
send attached dataset from KPACS to fodicom CSTORESCP. Even with fixing unhandled VR exception , whole study not getting sent to fodicom CSTORAGESCP SAMPLE. please make sure this issue gets resolved by making sure whole study well received by FODICOM CSTORAGESCP.

Error with Unhandled VR exception.. 

Error details..
OnCstoreRequestException occured Dicom.DicomFileException: Unhandled VR in DICOM parser observer: NONE ---> Dicom.DicomDataException: Unhandled VR in DICOM parser observer: NONE
at Dicom.IO.Reader.DicomDatasetReaderObserver.OnElement(IByteSource source, DicomTag tag, DicomVR vr, IByteBuffer data)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseValue(IByteSource source)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseDataset(IByteSource source)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseItemSequenceValue(IByteSource source)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseItemSequence(IByteSource source)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseValue(IByteSource source)
at Dicom.IO.Reader.DicomReader.DicomReaderWorker.ParseDataset(IByteSource source)
at Dicom.IO.Reader.DicomReader.Read(IByteSource source, IDicomReaderObserver observer, Func2 stop) at Dicom.IO.Reader.DicomFileReader.DoParse(IByteSource source, IDicomReaderObserver fileMetasetInfoObserver, IDicomReaderObserver datasetObserver, Func2 stop, DicomTransferSyntax& syntax, DicomFileFormat& fileFormat)
at Dicom.IO.Reader.DicomFileReader.Parse(IByteSource source, IDicomReaderObserver fileMetasetInfoObserver, IDicomReaderObserver datasetObserver, Func2 stop) at Dicom.IO.Reader.DicomFileReader.Read(IByteSource source, IDicomReaderObserver fileMetaInfo, IDicomReaderObserver dataset, Func2 stop)
at Dicom.DicomFile.Open(IFileReference file, Encoding fallbackEncoding)
--- End of inner exception stack trace ---
at Dicom.DicomFile.Open(IFileReference file, Encoding fallbackEncoding)
at Dicom.Network.DicomService.<ProcessPDataTFAsync>d__49.MoveNext()

### fo-dicom version and OS/platform
latest version beta4 , Windows Server 2012 R2

**Comment by gofal** [comment]
dglecons reported, the Exception occurs in DicomTag FFFE,E0DD. That is a special Tag for separating Items in a Sequence. This Tag never has a ValueRepresentation. Seems like a bug.
In http://dicom.nema.org/medical/Dicom/2014c/output/chtml/part05/sect_7.5.html there are descriptions of how to use the FFFE,E0DD or other Delimiter-Tags.

**Comment by gofal** [comment]
A fix could be the following:
in class DicomReader.cs in Methods ParseValue and ParseValueAsync replace the
`if (this._tag == DicomTag.ItemDelimitationItem)
{
// end of sequence item
return false;
}`

by the following:
`if (this._tag == DicomTag.ItemDelimitationItem || this._tag == DicomTag.SequenceDelimitationItem)
                {
                    // end of sequence item
                    return false;
                }`

**Comment by gofal** [comment]
Message from dglecons: 

> after that change, i was able to receive 5 more images After that i get rejection from KPACS

**Comment by gofal** [comment]
checked the files with dicomValidationToolkit. they are fine. The SequenceDelimitationItem is more precisely described here: http://dicom.nema.org/medical/Dicom/2016b/output/chtml/part05/sect_7.5.2.html
Dicom-dump looks like this:
```
(0008,1250) SQ (Sequence with undefined length #=1)     # u/l, 1 RelatedSeriesSequence
  (fffe,e000) na (Item with undefined length #=5)         # u/l, 1 Item
    (0020,0000) UL 140                                      #   4, 1 GenericGroupLength
    (0020,000d) UI [1.2.392.200036.9123.100.12.12.31013.90161209084010902810462310] #  62, 1 StudyInstanceUID
    (0020,000e) UI [1.2.392.200036.9123.100.12.12.31013.90161209084012906842910174] #  62, 1 SeriesInstanceUID
    (0040,0000) UL 20                                       #   4, 1 GenericGroupLength
    (0040,a170) SQ (Sequence with undefined length #=0)     # u/l, 1 PurposeOfReferenceCodeSequence
    (fffe,e0dd) na (SequenceDelimitationItem)               #   0, 0 SequenceDelimitationItem
  (fffe,e00d) na (ItemDelimitationItem)                   #   0, 0 ItemDelimitationItem
(fffe,e0dd) na (SequenceDelimitationItem)               #   0, 0 SequenceDelimitationItem
```
So the sequence als well as the Item have no length, but the FFFFFFFh value. that is why their end must be marked with a ItemDelimiatationItem or a SequenceDelimitationItem.

**Comment by anders9ustafsson** [comment]
Using the latest commit, I have no problems parsing and displaying the first or any arbitrary file in the dataset you sent, @dglecons. Could you please be more specific in which *file* and at which *tag* the exception is occurring? 

We did some fixes for parsing private sequence items in #346, not sure if that could have an impact on your problem?

**Comment by dglecons** [comment]
Have you tested while sending it from KPACS?

**Comment by anders9ustafsson** [comment]
@dglecons No, not yet. What happens when you read the file with the *DICOM Dump* application? And as I mentioned, it would help a lot if you could pinpoint exactly which file and at which tag the error occurs.

**Comment by anders9ustafsson** [comment]
BTW, are you using the sample as-is, @dglecons? Please note that the latest commit in *fo-dicom-samples*  uses *fo-dicom* version 3.0.0 Beta **1** from NuGet. Unless you have already done it previously, please update the NuGet package to Beta 4 and check the results.

(Please note that packaging has changed between Beta 1 and 4. I suggest that you first remove the existing NuGet packages from the C-Store SCP project, then add only the package *fo-dicom.Desktop* back to the project.)

**Comment by dglecons** [comment]
Already tested with beta4.Still the same error.Manually changed the version pointing to beta4 in the storagescp and tested.
Right now I am checking the problematic files and will send you soon.

**Comment by gofal** [comment]
I have also done some checks. I can open and parse every dicom-file. When sending the files with dcmtk (http://dicom.offis.de/dcmtk.php.de) to the fo-dicom server, everything works file. But when I send the study with k-pacs (downloaded version 1.6.0) then there are errors. It seems as it has to be with k-pacs.

**Comment by anders9ustafsson** [comment]
Thanks, @gofal, for checking! @dglecons, you might also want to take a look at #364, which exhibits the same error message.

**Comment by gofal** [comment]
My first assumption that there is a parsing error in fo-dicom was just from reading about the exception and reading in code. Now after I have done lot of tests I am sure **it is a bug in k-pacs**.
All of the images can be opened and parsed correctly with DicomDump, which is included in fo-dicom. when sending the images with a free tool dcmtk to the storescp, everything works fine, but when sending with k-pacs, then there are parser-errors.
The exception happens when the StoreSCP wants to access the DicomTag 0020,000D then there is no DicomTag 0020,000D. This screenshot shows, that all Tags up to 0009,0000 have been parsed, but none of the next tags:
![image](https://cloud.githubusercontent.com/assets/6769246/21242097/f260eb62-c312-11e6-99a1-6b9b9b3db8fc.png)
The value of this last Tag 0009,0000 is weird:
![image](https://cloud.githubusercontent.com/assets/6769246/21242179/5778c2fe-c313-11e6-8450-85079fcf6237.png)
I ran DVTk, a dicom network sniffer, and k-pacs sent the following data:
```
[...]
# (0xFFFEE00D, 0) # ItemDelimiter 
# (0xFFFEE0DD, 0) # SequenceDelimiter
) # Related Series Sequence with length of: 0
(0x00090000,UN,"62000000") # 
(0x00090010,LO,"MMCPrivate") # Private Creator Group 0009
(0x00091048,UN,"4D5249204849502042494C41544552414C20572F4F205249544553482020") # 
(0x0009104E,UN,"48495020") # 
(0x0009104F,UN,"556E50726F7465637420") # 
(0x00100000,UL,0x0000006C) # 
(0x00100010,PN,"ANONKP0000") # Patient's Name
(0x00100020,LO,"RRY33699") # Patient ID
[...]
(0x00101010,AS,"062Y") # Patient's Age
(0x00101030,DS,"68.04 ") # Patient's Weight
(0x00101060,PN,"") # Patient's Mother's Birth Name
(0x00110000,UN,"3A000000") # 
(0x00110010,UN,"4D4D4350726976617465") # 
(0x00111001,UN,"46616C736520") # 
(0x00111002,UN,"556E50726F7465637420") # 
(0x00180000,UL,0x0000024C) # 
(0x00180010,LO,"off ") # Contrast/Bolus Agent
(0x00180015,CS,"HIP ") # Body Part Examined
[...]
(0x00189034,CS,"LINEAR") # Rectilinear Phase Encode Reordering
(0x00189093,US,0x000E) # Number of k-Space Trajectories
(0x00190000,UN,"9A010000") # 
(0x00190010,UN,"4D4D4350726976617465") # 
(0x00191001,UN,"") # 
(0x00191002,UN,"53414720") # 
[...]
```
Here one can see that the GroupLength of public Groups are well encoded as UL, but the GroupLengths of private Groups are UN and the strings are coded wrong. For example the correct value of 0009,0000 should be 0x62, but there it is 62000000 with then fo-dicom converts as 0x6200000 = 1.644.167.168. 
For comparison here is the network sniffing when exactly the same images are sent with dcmtk:
```
# (0xFFFEE00D, 0) # ItemDelimiter 
) # Related Series Sequence with length of: 0
(0x00090000,UL,0x00000062) # 
(0x00090010,LO,"MMCPrivate") # Private Creator Group 0009
(0x00091048,UN,"4D5249204849502042494C41544552414C20572F4F205249544553482020") # 
(0x0009104E,UN,"48495020") # 
(0x0009104F,UN,"556E50726F7465637420") # 
(0x00100000,UL,0x0000006C) # 
(0x00100010,PN,"ANONKP0000") # Patient's Name
(0x00100020,LO,"RRY33699") # Patient ID
[...]
(0x00101010,AS,"062Y") # Patient's Age
(0x00101030,DS,"68.04 ") # Patient's Weight
(0x00101060,PN,"") # Patient's Mother's Birth Name
(0x00110000,UL,0x0000003A) # 
(0x00110010,LO,"MMCPrivate") # 
(0x00111001,UN,"46616C736520") # 
(0x00111002,UN,"556E50726F7465637420") # 
(0x00180000,UL,0x0000024C) # 
(0x00180010,LO,"off ") # Contrast/Bolus Agent
(0x00180015,CS,"HIP ") # Body Part Examined
[...]
(0x00189034,CS,"LINEAR") # Rectilinear Phase Encode Reordering
(0x00189093,US,0x000E) # Number of k-Space Trajectories
(0x00190000,UL,0x0000019A) # 
(0x00190010,LO,"MMCPrivate") # 
(0x00191001,UN,"") # 
(0x00191002,UN,"53414720") # 
```
Here every Group Length is encoded correctly.

conclusion:
K-pacs definitly has a bug and encodes the Group Lengths for private groups wrong.
Is this issue now closed, or should we implement some kind of validation so that a wrong Group Length encoding will be ignored?

**Comment by dglecons** [comment]
KPACS sends the same study to eFilm fine..I am getting all the dicom data exception errors like yours.So it must be with wrong group length.

Is there a way to handle this as eFilm works fine with KPACS with same study.

**Comment by dglecons** [comment]
We have recently moved our implementation from mdcm to fodicom.Everything works except this.
mdcm toolkit version that we have works fine with KPACS and able to receive this study fine.
So its good to have this resolved in our fodicom version.Please help.

**Comment by gofal** [comment]
I have done some research about the Group Length. It is only one single Tag that posts how big the data of the following group is. the reason for this is, when searching in a dicomfile for a special dicomtag, then you dont have to iterate all of the Tags but you can "jump over the groups" with your Streamreader until you come to the group you are interested in. 
Since 2008 this Group Length was retired but of course there are still some implementations creating this Tag. In my opinion when parsing the Group Length then this Tag could be completly ignored. I would go so far and even not include it into the _items collection of DataSet. Because if you load a image and then change a property (eg DicomTag 0010,0010 PatientName for anonymizing reason) then you would have to recalculate the length of all Tags in group 0010,xxxx and update the value of the Group Length 0010,0000. I am not sure if this logic is included in fo-dicom. The more save way would be, to ignore this tag and after changing and rewriting the dataset there will not be such a Group Length any more.

@anders9ustafsson : Would this be OK? Then I would start implementing and submitting it.

**Comment by anders9ustafsson** [comment]
@gofal Recalculation of group lengths is included in *fo-dicom*, but admittedly I have not verified that it functions properly under all circumstances. 

If I understand correctly, there are still scenarios where group lengths are expected (command messages and meta information headers?), so full retirement is not applicable as I see it.

Ignoring it in reading would be OK I suppose (**anyone in the community against?**) but would that necessarily solve the "K-PACS" problem? You would still have to parse through to locate the next tag. Not sure how well that works with the current implementation.

With these caveats in mind, I still would appreciate very much if would give it a try to fix the parsing problem, @gofal, if you believe you are able to do it with a limited effort. Maybe you can publish a works-in-progress branch which could be continuously reviewed, since I believe this issue may very well concern others in the community?

**Comment by gofal** [comment]
Yes, you're right, in Group 0002,0000 for example the Group Length is still required. I would like to give it a try and implement the ignoring in reading.
I'm rather new in github. until now I have forked the fo-dicom and the fo-dicom-samples and did some changes (I'm improving the universal viewer and will soon start implementing the some volume classes so that it is possible to calculate MPR). If I do the changes and make a pull request, then there are automatically all other changes for MPR included? Should I create a new fork and then pull it into fo-dicom as well as into my first fork?

**Comment by gofal** [comment]
I think I have the solution now. At least the store from k-pacs works fine and store of all the other sample images worked fine as well. Would be great if other people try this as well.
The changed lines in code are very view. How can I contribute this best?

**Comment by dglecons** [comment]
I would like to try this in our test environment..where can I get the changes...

**Comment by gofal** [comment]
my result so far: There have been 2 Problems. The first has been caused by a corrupt dicomimage. When dglecons anonymized this file with k-pacs then this corrupt part has been repaired by k-pacs while recreating the dataset, but at this point k-pacs encoded the private Group Lengths as UN where it should have been encoded as UL.

The second problem has been solved yesterday. When dglecons testet my solution he found out that the anonymized files work well, but the original files still throw an error.

Now I also have solved the first original problem:
The reason was the SharedFunctionGroupSequence 5200,9229 which contains one Element MRReceiveCoilSequence (a nested sequence element). It is encoded with undefined length, that means that after that one Element there would have to be a ItemDelimitationItem and then finally a SequenceDelimitationItem. 
Now this file had a SequenceDelimitationItem directly after this one Element.
In the original version of fo-dicom before all my changes, the parser expected at the end of the one element a ItemDelimitationItem or another second element. Since there was no ItemDeleimitationItem the parser wanted to parse it as a new element. but there was actually a SequenceDelimitationItem which has no valid ValueRepresentation and therefore was not parsable and we had the error of the "Unhandled VR".
My very first change solved that in a way that the parser did not try any more to parse the SequenceDelimitationItem.
But there still was the problem that the parser expects the ItemDelimitationItem and then the SequenceDelimitationItem, but the ItemDelimitationItem is missing. So now I have done a change so that when the Parser looks for a ItemDelimitationItem but there is a SequenceDelimitationItem instead, then the parser now finishes the list of Items AND finishes the Sequence. So the parser works fine now with this corrupt file.
the changes are now in my fork gofal/fo-dicom. Dglecons will test this version.

**Comment by anders9ustafsson** [comment]
@gofal Many thanks for looking so closely into this issue!

Do I understand correctly that your fix is in [this commit](https://github.com/gofal/fo-dicom/commit/a0d88afa6d5e5165e7c82fe948571b50b312b84c)?

To facilitate merging the correction, I would recommend that you create a new branch from the *fo-dicom* organization's *develop* branch (latest commit), add the correction, and post a pull request from your new branch.

Many thanks in advance!

**Comment by anders9ustafsson** [comment]
@dglecons Any news on the outcome of these tests?
@gofal Is the fix included in [this](https://github.com/gofal/fo-dicom/commit/a0d88afa6d5e5165e7c82fe948571b50b312b84c) commit, as I assumed?

**Comment by gofal** [comment]
yes. all my changes are in this commit. they solve the two problems described above. There are still tests running, thats why I didn't create this branch until now.

**Comment by gofal** [comment]
I communicate with dglecons in a private gitter chat. His result so far: sending this corrupt study with k-pacs works now fine. but with a older version of k-pacs (0.9.9) there is still a problem. I could reproduce it: the first image is sent and parsed without problem but then instead of a second image therre is a abort of the connection and fo-dicom gets a IOException that the stream was closed by the client. Spent quite some time in searching for a reason, but could not find any. 
I did some improvements on fo-dicom, so that it is more flexible on invalid dicom-files, but I am afraid I am not able to solve this issue completely.
